internal/provider/resource_application.go

```diff
@@ -349,151 +545,195 @@ func resourceApplicationRead(ctx context.Context, d *schema.ResourceData, meta i
-		oldConfigMap := oldConfig.(map[string]interface{})
-		newConfigMap := newConfig.(map[string]interface{})
-		for k, v := range newConfigMap {
+	if !plan.Config.Equal(state.Config) {
+		planConfigMap := map[string]string{}
+		stateConfigMap := map[string]string{}
+		resp.Diagnostics.Append(plan.Config.ElementsAs(ctx, &planConfigMap, false)...)
+		resp.Diagnostics.Append(state.Config.ElementsAs(ctx, &stateConfigMap, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		for k, v := range planConfigMap {
 			// we've lost the type of the config value. We compare the string
 			// values.
-			oldEntry := fmt.Sprintf("%#v", oldConfigMap[k])
+			oldEntry := fmt.Sprintf("%#v", stateConfigMap[k])
 			newEntry := fmt.Sprintf("%#v", v)
 			if oldEntry != newEntry {
 				if updateApplicationInput.Config == nil {
 					// initialize just in case
-					updateApplicationInput.Config = make(map[string]interface{})
+					updateApplicationInput.Config = make(map[string]string)
 				}
 				updateApplicationInput.Config[k] = v
 			}
 		}
 	}
 
-	if d.HasChange("constraints") {
-		_, newConstraints := d.GetChange("constraints")
-		appConstraints, err := constraints.Parse(newConstraints.(string))
+	if !plan.Constraints.Equal(state.Constraints) {
+		appConstraints, err := constraints.Parse(plan.Constraints.ValueString())
 		if err != nil {
-			return diag.FromErr(err)
+			resp.Diagnostics.AddError("Conversion", fmt.Sprintf("Unable to parse plan constraints, got error: %s", err))
 		}
 		updateApplicationInput.Constraints = &appConstraints
 	}
 
-	err = client.Applications.UpdateApplication(&updateApplicationInput)
-	if err != nil {
-		return diag.FromErr(err)
+	if err = r.client.Applications.UpdateApplication(&updateApplicationInput); err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to update application resource, got error: %s", err))
+		return
 	}
 
-	return resourceApplicationRead(ctx, d, meta)
+	plan.ID = types.StringValue(newAppID(plan.ModelName.ValueString(), plan.ApplicationName.ValueString()))
+	r.trace("Updated", applicationResourceModelForLogging(ctx, &plan))
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
 }
 
 // computeExposeDeltas computes the differences between the previously
 // stored expose value and the current one. The valueSet argument is used
 // to indicate whether the value was already set or not in the latest
 // read of the plan.
-func computeExposeDeltas(oldExpose interface{}, newExpose interface{}, valueSet bool) (map[string]interface{}, []string) {
-	var old map[string]interface{} = nil
-	var new map[string]interface{} = nil
-
-	if oldExpose != nil {
-		aux := oldExpose.([]interface{})
+func computeExposeDeltas(stateExpose interface{}, planExpose interface{}, valueSet bool) (map[string]interface{}, []string) {
+	// TODO hml 2023-08-04
+	// Is there a better way to do this, now that we can deal
+	// in structures rather than maps of interfaces?
```

Comment:
yeah I think we could compare `stateExposeSlice[0]` and `planExposeSlice[0]` directly. Though I suppose we'll still need to `transformToMapStringInterface` because obviously juju api takes a map. It might at least save us from all the nil checks.

Though this can be done later, of course.

Reply:
They have already been compared directly at ln 626: `if !plan.Expose.Equal(state.Expose) {`. With different if/else we know that neither the state nor plan expose blocks are nil.